### ISSUE TITLE
Adaptec JBOD metadata incompatible with (overwrites) ZFS, causes data loss

### ISSUE THREAD

**Issue by gkkovacs** [zfsonlinux/zfs#7442]
We had a Proxmox 4.x server (ZFS 0.6.5.9), installed on 3x Samsung 500GB 850 EVO SSDs in RAIDZ1, that died recently. The disks were put in another box, connected to an Adaptec 6805E controller, disks initialized then added as JBODs, then each changed to Morphed JBOD to enable booting.

Needless to say the ZFS pool is not booting anymore:
`Command:/sbin/zpool import -N 'rpool'`
`Message: cannot import 'rpool': no such pool available`
`Error: 1`

Manually importing the pool also does not work. Even more frightening that zdb does not see any uberblocks:
`# zdb -lu /dev/sdc`
`failed to unpack label 0`
`failed to unpack label 1`
`failed to unpack label 2`
`failed to unpack label 3`

`# zdb -luv /dev/sdc`
`failed to unpack label 0`
`------------------------------------`
`LABEL 0`
`------------------------------------`
`    Uberblock[0] invalid`
`    Uberblock[1] invalid`
`    Uberblock[2] invalid`
`    Uberblock[3] invalid`
`    Uberblock[4] invalid`
`    Uberblock[5] invalid`
...and this goes on for all labels.

Partitions seem to be there (although GPT seems damaged):
`# gdisk -l /dev/sdc`

`Disk /dev/sdc: 976773168 sectors, 465.8 GiB`
`Logical sector size: 512 bytes`
`Disk identifier (GUID): ED0333A2-7D53-4E57-92C5-DD1388CA0CDC`
`Partition table holds up to 128 entries`
`First usable sector is 34, last usable sector is 976773134`
`Partitions will be aligned on 8-sector boundaries`
`Total free space is 0 sectors (0 bytes)`

`Number  Start (sector)    End (sector)  Size       Code  Name`
`   1              34            2047   1007.0 KiB  EF02`
`   2            2048       976756749   465.8 GiB   BF01  zfs`
`   9       976756750       976773134   8.0 MiB     BF07`

So it seems that initializing the disk with the Adaptec BIOS utility and/or creating and morphing JBODs destroys some important ZFS metadata. Pools are not available to be imported anymore, all data seems lost.

**My first question is: is it possible to fix the pools that are on these disks somehow?** The partitions seem to be untouched, so all data should be there.

Then I dug deeper: since we have a few more boxes where ZFS was installed on Adaptec Morphed JBODs initially, I checked them with zdb. Interestingly, zdb gives the exact same result on the disks of perfectly working ZFS pools: **failed to unpack label** and **Uberblock invalid** messages.

**So not only Adaptec array metadata is somehow killing ZFS pools if connected to such controller, pools initially installed on Adaptec JBODs, working perfectly are also not recognized by zdb.**

**Comment by beren12** [comment]
Umm, if you really initialized the disks then you erased them. I haven't tried this controller but a quick google shows it's not a good controller to use, or at least swap between... Have you tried putting them in the old box or one with a different controlled? Some controllers mess with sector sizes, like my seagate usb enclosure. For future info, any controller that can't be used in HBA mode is dangerous to use...

A quick google shows:
http://ask.microsemi.com/app/answers/detail/a_id/3251/~/can-drives-be-used-on-an-adaptec-raid-controller-but-not-be-used-as-part-of-a

https://forums.freenas.org/index.php?threads/need-some-help-with-adaptec-6805e.39198/

**Comment by kpande** [comment]
please use [the zfs-discuss mailing list or IRC](http://list.zfsonlinux.org) for support requests, unless you find a bug in ZFS that led to this condition.

**Comment by gkkovacs** [comment]
@kpande If you read my last to paragraphs, there clearly seems to be a issue with zdb: it can't read metadata from otherwise perfectly working disks/pools.

**Comment by kpande** [comment]
it really sounds like you erased the metadata using the RAID controller, and you're not looking at the correct partition using `zdb`. you are using the whole disk `sdc` when you should be using `sdcX`.

it's not a bug in ZFS, it's an issue with the hardware you used.

**Comment by gkkovacs** [comment]
Turns out I did not erase metadata, or any data. Only the GPT partition table got messed up somehow (probably thanks to Adaptec), so partitions we not active, therefore ZFS could not see the pools. Fixing them with gdisk restored the original partitions, the whole RAIDZ pool is there intact, now bootable again.

Thanks for pointing that out that zdb needs correct partitions as input, not disks. Helped me solve this nightmare. Sorry for mislabeling this as a ZFS issue.
